src/System.IO.FileSystem/src/System/IO/DirectoryInfo.cs

```diff
@@ -4,42 +4,40 @@
 
 using System.Collections.Generic;
 using System.Diagnostics;
-using System.Diagnostics.Contracts;
-using System.Security;
 
 namespace System.IO
 {
     public sealed partial class DirectoryInfo : FileSystemInfo
     {
+        private string _name;
+
         public DirectoryInfo(string path)
         {
-            if (path == null)
-                throw new ArgumentNullException(nameof(path));
-            Contract.EndContractBlock();
-
-            OriginalPath = PathHelpers.ShouldReviseDirectoryPathToCurrent(path) ? "." : path;
-            FullPath = Path.GetFullPath(path);
-            DisplayPath = GetDisplayName(OriginalPath);
+            Init(originalPath: PathHelpers.ShouldReviseDirectoryPathToCurrent(path) ? "." : path,
+                  fullPath: Path.GetFullPath(path),
+                  isNormalized: true);
         }
 
-        internal DirectoryInfo(string fullPath, string originalPath)
+        internal DirectoryInfo(string originalPath, string fullPath = null, string fileName = null, bool isNormalized = false)
         {
-            Debug.Assert(Path.IsPathRooted(fullPath), "fullPath must be fully qualified!");
-
-            // Fast path when we know a DirectoryInfo exists.
-            OriginalPath = originalPath ?? Path.GetFileName(fullPath);
-            FullPath = fullPath;
-            DisplayPath = GetDisplayName(OriginalPath);
+            Init(originalPath, fullPath, fileName, isNormalized);
         }
 
-        public override string Name
+        private void Init(string originalPath, string fullPath = null, string fileName = null, bool isNormalized = false)
         {
-            get
-            {
-                return GetDirName(FullPath);
-            }
+            Debug.Assert(!isNormalized || !PathInternal.IsPartiallyQualified(fullPath ?? originalPath), "should be fully qualified if normalized");
+
+            // Want to throw the original argument name
+            OriginalPath = originalPath ?? throw new ArgumentNullException("path");
+            FullPath = isNormalized ? fullPath ?? originalPath : Path.GetFullPath(fullPath ?? originalPath);
```

Comment:
How about storing `fullPath ?? originalPath` into a local?  It's used three times in the above three lines.

Reply:
Ok